### PR TITLE
Add helpers for setting common configuration options

### PR DESCRIPTION
**Pull request by shalvah-gs**
This PR is part of the effort to make NxtHttpClient easier to use, this time by reducing the boilerplate involed in configuring a client. This PR takes this:

```rb
class Client < NxtHttpClient::Client
  configure do |config|
    config.base_url = url
    config.request_options.deep_merge!(
      headers: { 
        'Content-Type': 'application/json', 
        "Accept": 'application/json',
        'Authorization': 'Basic ' + Base64.strict_encode64("username:password")
      },
      timeout: 5
    )
  end

  response_handler do |handler|
    handler.on(:success) do |response|
      JSON(response.body)
    end

    handler.on(:error) do |response|
      error = NxtHttpClient::Error.new(response)
      Sentry.set_extras(error_details: error.to_h)

      raise error
    end
  end
end


client.post('/', {some: 'data'}.to_json)
```

and simplifies it into:

```rb
class Client < NxtHttpClient::Client
  configure do |config|
    config.base_url = url
    config.basic_auth = { username: 'username', password: 'password' }
    config.timeout_seconds total: 5

    config.json_request = true
    config.json_response = true
    config.raise_response_errors = true
  end
end

client.post('/', {some: 'data'})
```

Why? These are common options that we often need for API clients. They shouldn't be complicated or difficult to remember.

Of course, the old options are still available.

[BE-46][]

[BE-46]: https://hellogetsafe.atlassian.net/browse/BE-46?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/nxt_http_client/client.rb [hunk 1]
```diff
@@ -122,6 +140,26 @@ def callback_or_response(response, response_handler)
...
+            response
+          end
+        end
+      end
+
+      if config.raise_response_errors
+        response_handler.configure do |handler|
+          handler.on(:error) do |response|
+            error = NxtHttpClient::Error.new(response)
+            ::Sentry.set_extras(error_details: error.to_h) if defined?(::Sentry)
```
Comment (shalvah-gs) [inline lib/nxt_http_client/client.rb:157]: I'm not a fan of blindly sending request and response contents to Sentry, but this was the previous behaviour, so I've kept it. Maybe we could expose a filter proc or something later (although Sentry already has some filters built in).

File: README.md [hunk 2]
```diff
@@ -17,102 +16,112 @@ gem 'nxt_http_client'
...
+Register your default request options on the class level. Available options are:
+- `request_options`, passed directly to the underlying Typhoeus Request
+- `base_url=`
+- `x_request_id_proc=`
+- `json_request=`: Shorthand to set the Content-Type request header to JSON and automatically convert request bodies to JSON
+- `json_response=`: Shorthand to set the Accept request header and automatically convert success response bodies to JSON
+- `raise_response_errors=`: Makes the client raise a `NxtHttpClient::Error` for a non-success response. 
+  You can also do this manually by setting a response_handler.
+- `bearer_auth=`: Set a bearer token to be sent in the Authorization header
+- `basic_auth=`: Pass an array containing username and password, to be sent as Basic credentials in the Authorization header
```
Comment (AkihikoITOH) [inline README.md:123]: I'd prefer (1) `config.basic_auth = { username:, password: }` or (2) `config.basic_auth = [username, password]` over (3) `config.basic_auth = username, password`.
For one, I find the 3 a bit confusing for the same reason I find a method call with multiple arguments without parenthesis confusing (and this technically is one). And the (1) makes it crystal clear which one is which, although, it should be usually clear without keywords.

**Review by AkihikoITOH — APPROVED** [review]
I personally like these changes, but it feels like it goes against the recent tendency in the team to prefer explicitness & verboseness over "Ruby-magic" & convenience 🤔 

Perhaps bring it up in the backend channel or the next weekly?

File: README.md (hunk 2 above)
Comment (shalvah-gs) [reply README.md:123]: Sure, I can switch to (1).

Comment (shalvah-gs) [reply README.md]: Changed to a Hash: https://github.com/nxt-insurance/nxt_http_client/pull/158/commits/895317d0ed7582ca1a6bf0219b4e3a4293e7eb1c

**Comment by shalvah-gs** [comment]
> Perhaps bring it up in the backend channel or the next weekly?

Yeah, perhaps. I think this is "good" convenience, but I can see a few changes might not be universally welcome.

File: lib/nxt_http_client/client.rb (hunk 1 above)
Comment (lutfidemirci) [reply lib/nxt_http_client/client.rb:159]: Do you think error_details is bit too generic? Should we call it `http_error_details`?


Some extra  thought: Sentry breadcrumbs can be also interesting to use.

https://docs.sentry.io/product/issues/issue-details/breadcrumbs/

**Review by lutfidemirci — APPROVED** [review]
LOVE IT 😍

File: lib/nxt_http_client/client.rb [hunk 3]
```diff
@@ -74,10 +82,22 @@ def build_headers(opts)
...
       opts[:headers] ||= {}
 
-      if config.x_request_id_proc
-        opts[:headers][XRequestId] ||= config.x_request_id_proc.call
+      opts[:headers]['Content-Type'] ||= ApplicationJson if config.json_request
+      opts[:headers]['Accept'] ||= ApplicationJson if config.json_response
+
+      if config.basic_auth
+        begin
+          config.basic_auth => { username:, password: }
```
Comment (AkihikoITOH) [inline lib/nxt_http_client/client.rb:90]: Hold on, what's this syntax? Is it something new? (I'm talking about the `=>`)

Comment (AkihikoITOH) [reply lib/nxt_http_client/client.rb:90]: Oh, it's rightward assignment.

Comment (shalvah-gs) [reply lib/nxt_http_client/client.rb:90]: Yeah. This specific usage is colloquially called "rightward assignment", but it's an offshoot of Ruby 3's pattern matching (good explainer [here](https://rubyreferences.github.io/rubychanges/3.0.html#one-line-pattern-matching-with-)). You can use it for rightward assignment, if you will (`5 => x`), but more practically, for hash destructuring:

```rb
config.basic_auth => { username:, password: }

# same as
username = config.basic_auth[:username]
password = config.basic_auth[:password]
```

File: lib/nxt_http_client/client.rb (hunk 1 above)
Comment (shalvah-gs) [reply lib/nxt_http_client/client.rb:159]: Good point, it is indeed too generic. I've found it confusing in the past. `http_error_details` is better, thanks.

breadcrumbs are useful, but they're thin–the http breadcrumb just lets you state that a request happened, and what it's status code was, but not much more (headers, body).

File: lib/nxt_http_client/client.rb (hunk 3 above)
Comment (alecspopa) [reply lib/nxt_http_client/client.rb:90]: You learn something every day :)

**Comment by alecspopa** [comment]
Is the PR description wrong?
This is a bit confusing even though I saw the related comments: `config.basic_auth = username, password`. What is `username` and `password`?

**Comment by shalvah-gs** [comment]
> Is the PR description wrong?
> This is a bit confusing even though I saw the related comments: `config.basic_auth = username, password`. What is `username` and `password`?

My bad, I forgot to update it. It should be a hash (`config.basic_auth = { username: 'something', password: 'otherthing' }`)

**Pull request merged**